### PR TITLE
feat(PE-8388): add undername ownership

### PR DESCRIPTION
**Pull request by atticusofsparta**
## Summary by CodeRabbit

- New Features
  - Per-record ownership for undernames with richer metadata (display name, logo, description, keywords) and a Transfer-Record action with owner notifications.

- Improvements
  - Stricter validation and enforced limits for names, descriptions, keywords, addresses and JSON inputs.
  - Permission model refined to distinguish ANT-level vs record-level permissions; per-record permission checks added.
  - Public actions renamed for clarity (primary-name flows).

- Documentation
  - Added detailed design doc describing hierarchical undername ownership and flows.

- Tests
  - Extensive unit and end-to-end coverage for ownership, permissions, metadata, and edge cases.

- Chores
  - Updated ignore/exclude patterns; some tests marked skipped.

### CODE REVIEW

File: src/common/constants.lua [hunk 1]
```diff
@@ -13,4 +13,10 @@ constants.INVALID_TTL_MESSAGE = "Invalid TTL. TLL must be an integer between "
 	.. constants.MAX_TTL_SECONDS
 	.. " seconds"
 
+-- Shared metadata constants (used for both ANT and undername metadata)
+constants.MAX_NAME_LENGTH = 50
```
Comment (atticusofsparta) [inline src/common/constants.lua]: 61 is the max for undernames

File: src/common/main.lua [hunk 2]
```diff
@@ -188,16 +191,110 @@ function ant.init()
...
+		end
 
-		local name, transactionId, ttlSeconds, priority =
-			string.lower(msg["Sub-Domain"]),
-			msg["Transaction-Id"],
-			tonumber(msg["TTL-Seconds"]),
-			tonumber(msg["Priority"])
+		local transactionId = msg["Transaction-Id"]
+		local ttlSeconds = tonumber(msg["TTL-Seconds"])
+		local priority = tonumber(msg["Priority"])
```
Comment (atticusofsparta) [inline src/common/main.lua]: assert NO priority if undername owner is the caller.

File: src/common/main.lua [hunk 3]
```diff
@@ -188,16 +191,110 @@ function ant.init()
...
+			utils.validateKeywords(decodedKeywords)
+			keywords = decodedKeywords
+		end
+
 		collectgarbage()
-		return records.setRecord(name, transactionId, ttlSeconds, priority)
+		return records.setRecord(name, transactionId, ttlSeconds, priority, owner, recordName, logo, description, keywords)
+	end)
+
+	createActionHandler(ActionMap.SetRecordMetadata, function(msg)
```
Comment (atticusofsparta) [inline src/common/main.lua]: SetRecord is sufficient, remove setRecordMetadata

File: src/common/main.lua [hunk 4]
```diff
@@ -241,6 +338,82 @@ function ant.init()
 		return balances.setLogo(msg.Logo)
 	end)
 
+	createActionHandler(ActionMap.TransferRecordOwnership, function(msg)
+		local subdomain = string.lower(msg.Tags["Sub-Domain"])
+		local newOwner = msg.Tags["New-Owner"]
```
Comment (atticusofsparta) [inline src/common/main.lua]: Recipient

File: src/common/main.lua [hunk 5]
```diff
@@ -241,6 +338,82 @@ function ant.init()
...
+		assert(record ~= nil, "Record does not exist")
+		assert(record.owner ~= nil, "Record has no owner")
+
+		-- Check permissions (ANT owner/controllers can transfer any record, record owners can transfer their own)
+		utils.assertHasRecordPermission(caller, subdomain)
+
+		-- Use existing transfer function with proper garbage collection
+		collectgarbage("stop")
+		local result = records.transferRecordOwnership(subdomain, newOwner, msg.Tags["Allow-Unsafe-Addresses"])
+		collectgarbage("restart")
```
Comment (atticusofsparta) [inline src/common/main.lua]: can remove these collectgarbage calls

File: src/common/main.lua [hunk 6]
```diff
@@ -241,6 +338,82 @@ function ant.init()
...
+			Data = json.encode(result)
+		})
+
+		-- Send response back to caller
+		utils.Send(msg, {
+			Target = msg.From,
+			Action = "Record-Ownership-Transferred",
+			Data = json.encode(result)
+		})
+	end)
```
Comment (atticusofsparta) [inline src/common/main.lua]: workshop the names of the notices

File: src/common/main.lua [hunk 7]
```diff
@@ -241,6 +338,82 @@ function ant.init()
...
+
+	createActionHandler(ActionMap.RevokeRecordOwnership, function(msg)
```
Comment (atticusofsparta) [inline src/common/main.lua]: SetRecord covers this, remove

File: src/common/main.lua [hunk 8]
```diff
@@ -301,16 +474,33 @@ function ant.init()
...
+			local underscorePos = string.find(name, "_")
+			if underscorePos and underscorePos > 1 then
+				local subdomain = string.sub(name, 1, underscorePos - 1)
+				local record = Records[subdomain]
+				-- CRITICAL: Record owner can only set primary name for themselves
+				if record and record.owner == caller and recipient == caller then
+					isAuthorized = true
+				end
+			end
+		end
```
Comment (atticusofsparta) [inline src/common/main.lua]: plz unit test this

Comment (atticusofsparta) [reply src/common/main.lua]: Refer to Approve-Primary-Name utils in network process for util and unit tests

File: src/common/main.lua [hunk 9]
```diff
@@ -301,16 +474,33 @@ function ant.init()
...
+			-- Check if caller owns a subdomain AND is setting it for themselves
+			-- Names follow pattern: subdomain_antname
+			local underscorePos = string.find(name, "_")
+			if underscorePos and underscorePos > 1 then
+				local subdomain = string.sub(name, 1, underscorePos - 1)
+				local record = Records[subdomain]
+				-- CRITICAL: Record owner can only set primary name for themselves
+				if record and record.owner == caller and recipient == caller then
+					isAuthorized = true
+				end
```
Comment (atticusofsparta) [inline src/common/main.lua]: leverage assertPermissionForRecord util

File: src/common/main.lua [hunk 10]
```diff
@@ -322,15 +512,36 @@ function ant.init()
...
+				local underscorePos = string.find(lowerName, "_")
+				if underscorePos and underscorePos > 1 then
+					local subdomain = string.sub(lowerName, 1, underscorePos - 1)
+					local record = Records[subdomain]
+					-- Record owner can only remove their own subdomain primary name
+					if record and record.owner == caller then
+						isAuthorized = true
+					end
+				end
+			end
```
Comment (atticusofsparta) [inline src/common/main.lua]: refer to other comment re unit test for parsing record name and assertHasRecordPermission

File: src/common/records.lua [hunk 11]
```diff
@@ -37,14 +42,16 @@ function records.setRecord(name, transactionId, ttlSeconds, priority)
 		transactionId = transactionId,
 		ttlSeconds = ttlSeconds,
 		priority = name == "@" and 0 or priority,
+		-- Add optional fields only if provided
+		owner = owner,
+		name = recordName,
```
Comment (atticusofsparta) [inline src/common/records.lua]: lets go with displayName here and update other metadata references to match

File: src/common/records.lua [hunk 12]
```diff
@@ -37,14 +42,16 @@ function records.setRecord(name, transactionId, ttlSeconds, priority)
...
+		logo = logo,
+		description = description,
+		keywords = keywords,
 	}
 	collectgarbage("restart")
```
Comment (atticusofsparta) [inline src/common/records.lua]: remove collectgarbage call

File: src/common/records.lua [hunk 13]
```diff
@@ -79,4 +86,77 @@ function records.getRecords()
 	return antRecords
 end
 
+--- Transfer ownership of a record to a new owner
+---@param name string The name of the record
+---@param newOwner string The new owner address
+---@param allowUnsafeAddresses boolean|nil Whether to allow unsafe addresses
+---@return table Transfer details
+function records.transferRecordOwnership(name, newOwner, allowUnsafeAddresses)
```
Comment (atticusofsparta) [inline src/common/records.lua]: recipient to match handler var

File: src/common/records.lua [hunk 14]
```diff
@@ -79,4 +86,77 @@ function records.getRecords()
...
+
+	local previousOwner = Records[name].owner
+	Records[name].owner = nil
+
+	return {
+		subdomain = name,
+		previousOwner = previousOwner,
+		revoked = true
+	}
+end
```
Comment (atticusofsparta) [inline src/common/records.lua]: getting rid of this in favour of setRecord

File: src/common/records.lua [hunk 15]
```diff
@@ -79,4 +86,77 @@ function records.getRecords()
...
+		record.description = description
+	end
+	if keywords ~= nil then
+		record.keywords = keywords
+	end
+	collectgarbage("restart")
+	
+	return record
+end
+
```
Comment (atticusofsparta) [inline src/common/records.lua]: get rid of this in favour for setRecord

File: src/common/types.lua [hunk 16]
```diff
@@ -30,7 +30,12 @@
 ---@alias Record {
 --- transactionId: string,
 --- ttlSeconds: integer,
---- priority: integer|nil
+--- priority: integer|nil,
+--- owner: string|nil,
+--- name: string|nil,
```
Comment (atticusofsparta) [inline src/common/types.lua]: displayName

File: src/common/utils.lua [hunk 17]
```diff
@@ -194,6 +194,33 @@ function utils.assertHasPermission(from)
...
+function utils.assertHasRecordPermission(from, subdomain)
+	-- Check if has ANT-level permission first (owner or controller)
+	for _, c in ipairs(Controllers) do
+		if c == from then
+			return
+		end
+	end
+	if Owner == from or Balances[from] == 1 or ao.env.Process.Id == from then
+		return
+	end
```
Comment (atticusofsparta) [inline src/common/utils.lua:213]: this is the same as validateOwner

File: docs/madr/2-undername-ownership.md [hunk 18]
```diff
@@ -0,0 +1,259 @@
+# Undername Ownership - Delegated Control for ANT Records
```
Comment (atticusofsparta) [inline docs/madr/2-undername-ownership.md:1]: To update with comments and architecture changes

**Review by atticusofsparta — COMMENTED** [review]
After addressing comments please update tests as well

File: madr/undername-ownership.md
Comment (dtfiedler) [inline madr/undername-ownership.md:1]: why do we have two of the same doc in different locations?

File: spec/utils_spec.lua [hunk 19]
```diff
@@ -146,3 +146,164 @@ describe("utils.validateUndername", function()
...
+end)
+
+describe("undernameForName", function()
+	it("should get the undername name for a name with an undername", function()
+		local undername = "undername"
+		local basename = "basename"
+		local name = undername .. "_" .. basename
+
+		local undernameFromName = utils.undernameForName(name)
+		assert.are.same(undernameFromName, undername)
```
Comment (dtfiedler) [inline spec/utils_spec.lua:301]: 🤔 - what is the intent of this function? base names wouldn't be known by ANTs

File: src/common/balances.lua [hunk 20]
```diff
@@ -42,6 +43,7 @@ end
 ---@return table<string, string>
 function balances.setName(name)
 	assert(type(name) == "string", "Name must be a string")
+	assert(#name <= constants.MAX_NAME_LENGTH, "Name must not be longer than " .. constants.MAX_NAME_LENGTH .. " characters")
```
Comment (dtfiedler) [inline src/common/balances.lua:46]: question: why is this function on `balances`?

File: src/common/balances.lua [hunk 21]
```diff
@@ -60,7 +62,7 @@ end
 ---@return table<string, string>
 function balances.setDescription(description)
 	assert(type(description) == "string", "Description must be a string")
-	assert(#description <= 512, "Description must not be longer than 512 characters")
+	assert(#description <= constants.MAX_DESCRIPTION_LENGTH, "Description must not be longer than " .. constants.MAX_DESCRIPTION_LENGTH .. " characters")
```
Comment (dtfiedler) [inline src/common/balances.lua:65]: same question here

File: src/common/main.lua [hunk 22]
```diff
@@ -64,6 +64,8 @@ function ant.init()
 		SetDescription = "Set-Description",
 		SetKeywords = "Set-Keywords",
 		SetLogo = "Set-Logo",
+		TransferRecordOwnership = "Transfer-Record-Ownership",
```
Comment (dtfiedler) [inline src/common/main.lua]: suggestion: `TransferRecord`

File: src/common/main.lua [hunk 23]
```diff
@@ -188,16 +190,71 @@ function ant.init()
...
-		utils.assertHasPermission(msg.From)
-
-		local name = string.lower(msg.Tags["Sub-Domain"])
-		local transactionId = msg.Tags["Transaction-Id"]
-		local ttlSeconds = tonumber(msg.Tags["TTL-Seconds"])
-		local priority = tonumber(msg.Tags["Priority"])
+		local name = string.lower(msg["Sub-Domain"])
+		local transactionId = msg["Transaction-Id"]
+		local ttlSeconds = tonumber(msg["TTL-Seconds"])
+		local priority = tonumber(msg["Priority"])
```
Comment (dtfiedler) [inline src/common/main.lua]: question: what are the ramifications of undername owners being able to set priority? what if there are collisions?

suggestion: `priority` should be a protected field that can only be modified by controllers/owners

File: src/common/main.lua [hunk 24]
```diff
@@ -188,16 +190,71 @@ function ant.init()
...
+		end
 
 		assert(ttlSeconds, "Missing ttl seconds")
-		collectgarbage()
-		return records.setRecord(name, transactionId, ttlSeconds, priority)
+
+		-- Handle optional metadata fields
+
+		-- Owner assignment requires ANT-level permission (only when explicitly setting a new owner)
+		local explicitOwner = msg.Tags["Owner"]
```
Comment (dtfiedler) [inline src/common/main.lua]: suggestion: be consistent and use `Recipient`

File: src/common/main.lua [hunk 25]
```diff
@@ -241,6 +298,43 @@ function ant.init()
...
 	end)
 
+	createActionHandler(ActionMap.TransferRecordOwnership, function(msg)
+		local subdomain = string.lower(msg.Tags["Sub-Domain"])
+		local recipient = msg.Tags["Recipient"]
+		local caller = msg.From
+
+		-- Validate inputs
+		assert(subdomain, "Sub-Domain is required")
+		assert(recipient, "Recipient is required")
```
Comment (dtfiedler) [inline src/common/main.lua]: suggestion: assert it is a valid address or that allow-unsafe-address is true

File: madr/undername-ownership.md
Comment (atticusofsparta) [reply madr/undername-ownership.md:1]: not sure, i think maybe i didnt realize one existed (probably docs)

File: spec/utils_spec.lua (hunk 19 above)
Comment (atticusofsparta) [reply spec/utils_spec.lua:301]: the primary name workflows get called with the arns "fqdn" so it comes with the base name attached. We dont particularly care about the base name, but we need to extract the undername

File: src/common/balances.lua (hunk 20 above)
Comment (atticusofsparta) [reply src/common/balances.lua:46]: I think the idea was that balances was to have the token spec. It more belongs in like.... "metadata" or maybe an "info" file. Or its own file. Decision to put it here was made long ago and I honestly don't remember exactly why.

File: src/common/balances.lua (hunk 21 above)
Comment (atticusofsparta) [reply src/common/balances.lua:65]: Well for this one we were following the previous pattern of having the metadata utils here.

File: src/common/main.lua (hunk 23 above)
Comment (atticusofsparta) [reply src/common/main.lua]: If priority is nil we validate that they are having record permission, if its not nil we verify that they are the owner/controllers of the ant (see below, ln:205)

You are right to question the impact there. If undername owners can set the priority they can undercut other domains and affect how others are served. Its probably not a huge deal but we decided it was not worth having undername owners have the ability to set it.

File: test/egress-json-memory-limits-test.mjs
Comment (dtfiedler) [inline test/egress-json-memory-limits-test.mjs:1]: was this intended for this PR?

File: docs/madr/2-undername-ownership.md [hunk 26]
```diff
@@ -0,0 +1,259 @@
...
+
+Records now support optional ownership and metadata fields:
+
+```lua
+---@alias Record {
+--- transactionId: string,
+--- ttlSeconds: integer,
+--- priority: integer|nil,
+--- owner: string|nil,           -- NEW: Record owner address
+--- name: string|nil,             -- NEW: Display name (max 50 chars)
```
Comment (dtfiedler) [inline docs/madr/2-undername-ownership.md:62]:
```suggestion
--- displayName: string|nil,             -- NEW: Display name (max 50 chars)
```

File: docs/madr/2-undername-ownership.md [hunk 27]
```diff
@@ -0,0 +1,259 @@
...
+
+### New Handlers
+
+#### Transfer-Record-Ownership
+Transfers ownership of a specific record to a new address:
+```lua
+{
+  Action = "Transfer-Record-Ownership",
+  SubDomain = "example",
+  NewOwner = "address123..."
```
Comment (dtfiedler) [inline docs/madr/2-undername-ownership.md]:
```suggestion
  Recipient = "address123..."
```

File: src/common/main.lua [hunk 28]
```diff
@@ -190,11 +190,12 @@ function ant.init()
 	end)
 
 	createActionHandler(ActionMap.SetRecord, function(msg)
+		print("SetRecord", json.encode(msg.Tags))
```
Comment (dtfiedler) [inline src/common/main.lua]:
```suggestion
```

File: docs/madr/2-undername-ownership.md [hunk 29]
```diff
@@ -0,0 +1,296 @@
...
+
+1. **ANT Owner**: Full control via ownership (balance = 1) or process ID
+2. **Controllers**: Full control via Controllers array
+3. **Record Owners**: Control only their specific record
+
+### New Handlers
+
+#### Transfer-Record
+
+Transfers ownership of a specific record to a new address:
```
Comment (dtfiedler) [inline docs/madr/2-undername-ownership.md]:
```suggestion
Transfers or sets ownership of a specific record to a new address:
```

File: docs/madr/2-undername-ownership.md [hunk 30]
```diff
@@ -0,0 +1,296 @@
...
+
+```lua
+{
+  Action = "Transfer-Record",
+  SubDomain = "example",
+  NewOwner = "address123..."
```
Comment (dtfiedler) [inline docs/madr/2-undername-ownership.md]:
```suggestion
  Recipient = "address123..."
```

File: docs/madr/2-undername-ownership.md (hunk 29 above)
Comment (dtfiedler) [reply docs/madr/2-undername-ownership.md]: disregard - an owner would need to set the `Record-Owner` via `Set-Record`

File: src/common/main.lua [hunk 31]
```diff
@@ -241,6 +298,43 @@ function ant.init()
...
+		local caller = msg.From
+
+		-- Validate inputs
+		assert(subdomain, "Sub-Domain is required")
+		assert(utils.isValidAOAddress(recipient, msg.Tags["Allow-Unsafe-Addresses"]), "Invalid recipient address")
+
+		-- Check if record exists and has an owner
+		local record = Records[subdomain]
+		assert(record ~= nil, "Record does not exist")
+		assert(record.owner ~= nil, "Record has no owner")
```
Comment (dtfiedler) [inline src/common/main.lua]: nit: consider adding more detailed error indicating a record must have an owner to be transferred to another owner

File: src/common/main.lua [hunk 32]
```diff
@@ -241,6 +298,43 @@ function ant.init()
...
+
+		-- Check permissions (ANT owner/controllers can transfer any record, record owners can transfer their own)
+		utils.assertHasRecordPermission(caller, subdomain)
```
Comment (dtfiedler) [inline src/common/main.lua:278]: controllers have ability to set/change owners of records.

@atticusofsparta - ensure spec is updated accordingly

File: src/common/main.lua [hunk 33]
```diff
@@ -241,6 +298,43 @@ function ant.init()
...
+
+		-- Use existing transfer function with proper garbage collection
+		local result = records.transferRecord(subdomain, recipient, msg.Tags["Allow-Unsafe-Addresses"])
```
Comment (dtfiedler) [inline src/common/main.lua]: nit: duplicate assertions in these function calls

File: src/common/main.lua [hunk 34]
```diff
@@ -241,6 +298,43 @@ function ant.init()
...
+
+		-- Send ownership transfer notice to new owner
+		ao.send({
+			Target = recipient,
+			Action = "Record-Ownership-Transfer-Notice",
```
Comment (dtfiedler) [inline src/common/main.lua]:
```suggestion
			Action = "Transfer-Record-Notice",
```

File: src/common/main.lua [hunk 35]
```diff
@@ -241,6 +298,43 @@ function ant.init()
...
+			["Previous-Owner"] = result.previousOwner,
+			Data = json.encode(result),
+		})
+
+		-- Send response back to caller
+		utils.Send(msg, {
+			Target = msg.From,
+			Action = "Record-Ownership-Transferred",
+			Data = json.encode(result),
+		})
```
Comment (dtfiedler) [inline src/common/main.lua]:
```suggestion
		return result; -- sends a `Transfer-Record-Notice` to the caller
```

File: src/common/records.lua [hunk 36]
```diff
@@ -79,4 +113,26 @@ function records.getRecords()
...
+	assert(recipient ~= Records[name].owner, "New owner same as current owner")
+
+	local previousOwner = Records[name].owner
+	Records[name].owner = recipient
+
+	return {
+		subdomain = name,
+		previousOwner = previousOwner,
+		recipient = recipient,
+	}
```
Comment (dtfiedler) [inline src/common/records.lua]:
```suggestion
	return {
		subdomain = name,
		previousOwner = previousOwner,
		newOwner = recipient,
		record = Records[name],
	}
```

File: docs/madr/2-undername-ownership.md [hunk 37]
```diff
@@ -0,0 +1,296 @@
...
+#### Revoke-Record-Ownership
+
+ANT owner/controllers can revoke any record ownership:
+
+```lua
+{
+  Action = "Revoke-Record-Ownership",
+  SubDomain = "example"
+}
+```
```
Comment (dtfiedler) [inline docs/madr/2-undername-ownership.md]: remove as it would be `setRecord` with `Record-Owner=nil`

Comment (dtfiedler) [reply docs/madr/2-undername-ownership.md]: @atticusofsparta to add tests for removing the owner after setting

Comment (dtfiedler) [reply docs/madr/2-undername-ownership.md]: after discussion it may be best to have a handler that supports explicitly removing a record owner to avoid having to send some arbitrary value in tags (e.g. `Record-Owner = 'nil'`)

File: docs/madr/2-undername-ownership.md [hunk 38]
```diff
@@ -0,0 +1,296 @@
...
+{
+  Action = "Set-Record-Metadata",
+  SubDomain = "example",
+  Owner = "address123...",         -- Optional (ANT owner/controllers only)
+  ["Record-Name"] = "New Name",    -- Optional
+  ["Record-Logo"] = "logoTx123...",-- Optional
+  ["Record-Description"] = "...",  -- Optional
+  ["Record-Keywords"] = '["blog"]' -- Optional (JSON array)
+}
+```
```
Comment (dtfiedler) [inline docs/madr/2-undername-ownership.md]:
```suggestion
```

Comment (dtfiedler) [reply docs/madr/2-undername-ownership.md]: all of this is managed via `Set-Record`

File: docs/madr/2-undername-ownership.md [hunk 39]
```diff
@@ -0,0 +1,296 @@
...
+
+Now accepts optional ownership and metadata parameters:
+
+```lua
+{
+  Action = "Set-Record",
+  SubDomain = "example",
+  TransactionId = "tx123...",
+  TtlSeconds = "3600",
+  Owner = "address123...",      -- Optional
```
Comment (dtfiedler) [inline docs/madr/2-undername-ownership.md]:
```suggestion
  `Record-Owner` = "address123...",      -- Optional
```

File: docs/madr/2-undername-ownership.md [hunk 40]
```diff
@@ -0,0 +1,296 @@
...
+#### Set-Record
+
+Now accepts optional ownership and metadata parameters:
+
+```lua
+{
+  Action = "Set-Record",
+  SubDomain = "example",
+  TransactionId = "tx123...",
+  TtlSeconds = "3600",
```
Comment (dtfiedler) [inline docs/madr/2-undername-ownership.md]:
```suggestion
  TTL-Seconds = "3600",
```

File: docs/madr/2-undername-ownership.md [hunk 41]
```diff
@@ -0,0 +1,296 @@
...
+
+#### Set-Record
+
+Now accepts optional ownership and metadata parameters:
+
+```lua
+{
+  Action = "Set-Record",
+  SubDomain = "example",
+  TransactionId = "tx123...",
```
Comment (dtfiedler) [inline docs/madr/2-undername-ownership.md]:
```suggestion
  Transaction-Id = "tx123...",
```

File: docs/madr/2-undername-ownership.md [hunk 42]
```diff
@@ -0,0 +1,296 @@
...
+### Updated Handlers
+
+#### Set-Record
+
+Now accepts optional ownership and metadata parameters:
+
+```lua
+{
+  Action = "Set-Record",
+  SubDomain = "example",
```
Comment (dtfiedler) [inline docs/madr/2-undername-ownership.md]:
```suggestion
  Sub-Domain = "example",
```

File: docs/madr/2-undername-ownership.md [hunk 43]
```diff
@@ -0,0 +1,296 @@
...
+  Keywords = '["defi","dao"]'    -- Optional, JSON array
+}
+```
+
+#### Approve-Primary-Name / Remove-Primary-Names
+
+Record owners can only set/remove primary names for themselves:
+
+- Must be the record owner
+- Recipient must match the caller's address
```
Comment (dtfiedler) [inline docs/madr/2-undername-ownership.md]: TLDR - record owners can approve primary name requests for themselves

File: src/common/main.lua [hunk 44]
```diff
@@ -304,17 +398,24 @@ function ant.init()
...
+
+		local undername = utils.undernameForName(name)
+
+		if undername == nil then
+			-- Only ANT owner can approve the @ record
+			utils.assertHasPermission(caller)
+		else
+			utils.validateUndername(undername)
+			utils.assertHasRecordPermission(caller, undername)
+			assert(recipient == caller, "Undername owners can only approve names for themselves")
```
Comment (dtfiedler) [inline src/common/main.lua]: suggestion: wrap these assertions in a single util that asserts caller = recipient = owner

File: src/common/main.lua [hunk 45]
```diff
@@ -304,17 +398,24 @@ function ant.init()
 	end)
 
 	createActionHandler(ActionMap.ApproveName, function(msg)
```
Comment (dtfiedler) [inline src/common/main.lua]: suggestion:
- if the undername has an owner, you don't need to parse/provide the recipient tag - it must be from the ApproveName must be from `owner`

File: docs/madr/2-undername-ownership.md (hunk 37 above)
Comment (atticusofsparta) [reply docs/madr/2-undername-ownership.md]: After further thought we are shifting to requiring all record parameters to be passed in, so no separate handler needed.

File: src/common/main.lua (hunk 45 above)
Comment (atticusofsparta) [reply src/common/main.lua]: Decided for API simplicity to require recipient.

**Comment by atticusofsparta** [comment]
currently testing with wasm binary: `FKtQtOOtlcWCW2pXrwWFiCSlnuewMZOHCzhulVkyqBE`

File: docs/madr/2-undername-ownership.md [hunk 46]
```diff
@@ -0,0 +1,248 @@
...
+}
+```
+
+Permissions:
+
+- ANT owner OR corresponding record owner for each name
+
+### Security Measures
+
+1. **Permission Validation**: `assertHasPermission()` and
```
Comment (dtfiedler) [inline docs/madr/2-undername-ownership.md:172]: these appear to no longer be used

File: src/common/records.lua [hunk 47]
```diff
@@ -12,39 +13,85 @@ Records = Records
...
-	Records[name] = {
-		transactionId = transactionId,
-		ttlSeconds = ttlSeconds,
-		priority = name == "@" and 0 or priority,
-	}
-	collectgarbage("restart")
+	-- Validate @ record priority before creating the record
+	if name == "@" and priority ~= nil and priority ~= 0 then
+		error("Cannot assign priority to @ record")
+	end
```
Comment (dtfiedler) [inline src/common/records.lua]: suggestion: this can be an assertion, following existing patterns

File: src/common/records.lua [hunk 48]
```diff
@@ -12,39 +13,85 @@ Records = Records
...
 
-	return {
+	local newRecord = {
 		transactionId = transactionId,
-		ttlSeconds = ttlSeconds,
-		priority = priority,
+		ttlSeconds = tonumber(ttlSeconds),
```
Comment (dtfiedler) [inline src/common/records.lua]: the expectation of this function is `ttlSeconds` is already an integer, what's the reason for `tonumber` here?

Comment (dtfiedler) [reply src/common/records.lua]: same question for `priority`

File: src/common/records.lua [hunk 49]
```diff
@@ -12,39 +13,85 @@ Records = Records
...
+	assert(
+		newRecord.owner == nil or utils.isValidAOAddress(newRecord.owner, allowUnsafeAddresses),
+		"Invalid owner address"
+	)
+	assert(newRecord.displayName == nil or #newRecord.displayName <= constants.MAX_NAME_LENGTH, "Invalid display name")
+	assert(newRecord.logo == nil or utils.isValidArweaveAddress(newRecord.logo), "Invalid logo")
+	assert(
+		newRecord.description == nil or #newRecord.description <= constants.MAX_DESCRIPTION_LENGTH,
+		"Invalid description"
+	)
```
Comment (dtfiedler) [inline src/common/records.lua]: comment: from a readability standpoint, it seems to odd to create `newRecord` - then do the assertions on the values

File: test/record-ownership.test.mjs [hunk 50]
```diff
@@ -0,0 +1,1181 @@
...
+    // Verify record was created with ownership
+    const recordResult = await handle(
+      {
+        Tags: [
+          { name: 'Action', value: 'Record' },
+          { name: 'Sub-Domain', value: 'test' },
+        ],
+      },
+      setRecordResult.Memory,
+    );
```
Comment (dtfiedler) [inline test/record-ownership.test.mjs:72]: suggestion: use helper functions to call these handlers consistently across our tests `e.g. `getRecord()` `setRecord()` that return the result and memory.

File: test/record-ownership.test.mjs [hunk 51]
```diff
@@ -0,0 +1,1181 @@
...
+
+      assert(
+        transferNoOwnerResult.Messages[0].Tags.find((t) => t.name === 'Error'),
+        'Should error when record has no owner',
+      );
+      assert(
+        transferNoOwnerResult.Messages[0].Data.includes('Record has no owner'),
+        'Error should mention record has no owner',
+      );
+    });
```
Comment (dtfiedler) [inline test/record-ownership.test.mjs:344]: this is where we missed testing coverage - fetch the record again and assert the owner was not set from this action

File: test/record-ownership.test.mjs [hunk 52]
```diff
@@ -0,0 +1,1181 @@
...
+      assert(
+        transferInvalidResult.Messages[0].Tags.find((t) => t.name === 'Error'),
+        'Should error on invalid recipient address',
+      );
+      assert(
+        transferInvalidResult.Messages[0].Data.includes(
+          'Invalid recipient address',
+        ),
+        'Error should mention invalid address',
+      );
```
Comment (dtfiedler) [inline test/record-ownership.test.mjs:399]: same comment: fetch the record again and confirm the owner represented in state did not change

File: test/record-ownership.test.mjs [hunk 53]
```diff
@@ -0,0 +1,1181 @@
...
+
+      // Verify ownership notice was sent
+      const notice = antOwnerTransferResult.Messages.find(
+        (m) =>
+          m.Target === recipient &&
+          m.Tags.find(
+            (t) => t.name === 'Action' && t.value === 'Transfer-Record-Notice',
+          ),
+      );
+      assert(notice, 'Ownership transfer notice should be sent to new owner');
```
Comment (dtfiedler) [inline test/record-ownership.test.mjs:549]: suggestion: fetch record and assert owner did in-fact change

File: docs/madr/2-undername-ownership.md (hunk 46 above)
Comment (atticusofsparta) [reply docs/madr/2-undername-ownership.md:172]: see records.lua for usage in setRecord

File: src/common/records.lua (hunk 48 above)
Comment (atticusofsparta) [reply src/common/records.lua]: holdover - will update

File: test/record-ownership.test.mjs (hunk 50 above)
Comment (atticusofsparta) [reply test/record-ownership.test.mjs:72]: We do - something I would like to see here actually is something like the LocalANT class for the ario sdk write tests that mount the process in aoloader have the wrapped dryrun/message apis with our native sdk calls.

Obviously thats a bit chicken/egg here since this is where we build the apis but the idea is the same. A class that maintains memory and results with a simple `reset` method to set the start memory back.

Could even have a little pipe utility there to output and parse messages for a cranking workflow with the ario process module (would be a fixture).

Bit in the weeds there - thats what i want out of an "ao process testing library" though.

File: src/common/records.lua (hunk 47 above)
Comment (atticusofsparta) [reply src/common/records.lua]: https://github.com/ar-io/ar-io-ant-process/pull/119/commits/8ace515069e4077b227dbaf99edf054b689be5d9

File: src/common/records.lua (hunk 48 above)
Comment (atticusofsparta) [reply src/common/records.lua]: [967f16a](https://github.com/ar-io/ar-io-ant-process/pull/119/commits/967f16a8f9c42982ed6fc6bff9736f565ba6c089)

File: src/common/records.lua (hunk 49 above)
Comment (atticusofsparta) [reply src/common/records.lua]: https://github.com/ar-io/ar-io-ant-process/pull/119/commits/69f9bb67ffb3cdff8834cd6fdefe56efec529093

File: test/record-ownership.test.mjs (hunk 53 above)
Comment (atticusofsparta) [reply test/record-ownership.test.mjs:549]: https://github.com/ar-io/ar-io-ant-process/pull/119/commits/d9904c2f58f8378fc648bb64fa7787efb9e7a747

File: test/record-ownership.test.mjs (hunk 52 above)
Comment (atticusofsparta) [reply test/record-ownership.test.mjs:399]: https://github.com/ar-io/ar-io-ant-process/pull/119/commits/d9904c2f58f8378fc648bb64fa7787efb9e7a747

File: test/record-ownership.test.mjs (hunk 51 above)
Comment (atticusofsparta) [reply test/record-ownership.test.mjs:344]: https://github.com/ar-io/ar-io-ant-process/pull/119/commits/d9904c2f58f8378fc648bb64fa7787efb9e7a747

File: test/records.test.mjs [hunk 54]
```diff
@@ -217,18 +237,72 @@ describe('aos Records', async () => {
...
+    // Test with explicit priority 0
+    const setRecordResult1 = await handle({
+      Tags: [
+        { name: 'Action', value: 'Set-Record' },
+        { name: 'Sub-Domain', value: '@' },
+        { name: 'Transaction-Id', value: ''.padEnd(43, '4') },
+        { name: 'TTL-Seconds', value: 900 },
+        { name: 'Priority', value: 0 },
+      ],
+    });
```
Comment (dtfiedler) [inline test/records.test.mjs:297]: suggestion: create a `setRecord` util that handles setting these tags consistently and returns result and memory

**Review by dtfiedler — APPROVED** [review]
some comments on tests, but otherwise LG.
